### PR TITLE
Scheduler: Use PodInfo instead of Pod for nominatedPods and QueuedPodInfo

### LINKED ISSUE #97813: Scheduler: Reduce unnecessary PodInfo struct initialization
**What would you like to be added**:

Originated from @ahg-g's TODO `AddPod/RemovePod should pass a *framework.PodInfo type instead of *v1.Pod.`. Some discussions at https://github.com/kubernetes/kubernetes/pull/97599#discussion_r550214325.

In scheduler, we use `PodInfo` to wrap pod with additional pre-computed information to accelerate processing. This issue is trying to reduce the construction of unnecessary PodInfo.

I will break this task into below items to make this more clearer. Now we have several places use `NewPodInfo`:

- [x] `NodeInfo.AddPod(pod *v1.Pod)`, caller of this function may call some of below functions as well, so we can pass PodInfo to this function and reuse it from some the callers sometimes.
- [x] `preFilterState.updateWithPod` -> `InterPodAffinity.Add/RemovePod` -> `PluginsRunner.RunPreFilterExtensionAdd/RemovePod`-> `frameworkImpl.RunFilterPluginsWithNominatedPods` | `DefaultPreemption.FindCandidates` (skip some private functions in the path), in both of the final callers, they will call `NodeInfo.AddPod(pod *v1.Pod)` as well.

`NewPodInfo` will also be called in `InterPodAffinity.PreFilter` and `InterPodAffinity.PreScore`, they will be called in `genericScheduler.Schedule`, but InterPodAffinity may be disabled, so I think there is no need to change them.

- [ ] ~~`PodNominator.NominatedPodsForNode` do not call `NewPodInfo` directly, but it relates with both`frameworkImpl.RunFilterPluginsWithNominatedPods` and `DefaultPreemption.FindCandidates`. We can track it with PodInfos rather than pods~~

**Why is this needed**:

This can help with scheduler performance with reducing complicated object construction.

/cc @ahg-g  @alculquicondor @chendave 
/sig scheduling
/assign

### PR DESCRIPTION
**Pull request by gavinfish**
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

Instead of constructing podInfo when retrieving nominated pods, we can just store podInfo for `nominatedPodMap`.

#### Which issue(s) this PR fixes:

Fixes #97813

#### Does this PR introduce a user-facing change?

```release-note
none
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by ahg-g** [comment]
Can you embed PodInfo in QueuedPodInfo? we need this for https://github.com/kubernetes/kubernetes/pull/98446, so might as well just do it in this PR?

**Comment by gavinfish** [comment]
> Can you embed PodInfo in QueuedPodInfo ([replace Pod With PodInfo](https://github.com/kubernetes/kubernetes/blob/master/pkg/scheduler/framework/types.go#L47))? we need this for #98446, so might as well just do it in this PR?

Done. PTAL.

File: pkg/scheduler/internal/queue/scheduling_queue.go [hunk 1]
```diff
@@ -458,7 +458,7 @@ func (p *PriorityQueue) Update(oldPod, newPod *v1.Pod) error {
 	// If pod is not in any of the queues, we put it in the active queue.
 	err := p.activeQ.Add(p.newQueuedPodInfo(newPod))
 	if err == nil {
-		p.PodNominator.AddNominatedPod(newPod, "")
+		p.PodNominator.AddNominatedPod(framework.NewPodInfo(newPod), "")
```
Comment (ahg-g) [inline pkg/scheduler/internal/queue/scheduling_queue.go:461]: lets do this here so we avoid creating podInfo twice:
```
pInfo := p.newQueuedPodInfo(newPod)
if err := p.activeQ.Add(pInfo); err != nil {
 return err
}
p.PodNominator.AddNominatedPod(pInfo, "")
p.cond.Broadcast()
return nil
```

**Review by ahg-g — COMMENTED** [review]
Thanks, comments mostly on avoiding to call `NewPodInfo` frequently.

File: pkg/scheduler/framework/types.go [hunk 2]
```diff
@@ -44,7 +44,7 @@ var generation int64
 // the pod's status in the scheduling queue, such as the timestamp when
 // it's added to the queue.
 type QueuedPodInfo struct {
-	Pod *v1.Pod
+	PodInfo *PodInfo
```
[the remark this answers is not in the document]
Comment (gavinfish) [reply pkg/scheduler/framework/types.go]: done.

File: pkg/scheduler/internal/queue/scheduling_queue.go [hunk 3]
```diff
@@ -197,7 +197,7 @@ var _ SchedulingQueue = &PriorityQueue{}
 // newQueuedPodInfoNoTimestamp builds a QueuedPodInfo object without timestamp.
 func newQueuedPodInfoNoTimestamp(pod *v1.Pod) *framework.QueuedPodInfo {
```
[the remark this answers is not in the document]
Comment (gavinfish) [reply pkg/scheduler/internal/queue/scheduling_queue.go]: done.

File: pkg/scheduler/internal/queue/scheduling_queue.go [hunk 4]
```diff
@@ -262,7 +262,7 @@ func (p *PriorityQueue) Add(pod *v1.Pod) error {
 		klog.Errorf("Error: pod %v is already in the podBackoff queue.", nsNameForPod(pod))
 	}
 	metrics.SchedulerQueueIncomingPods.WithLabelValues("active", PodAdd).Inc()
-	p.PodNominator.AddNominatedPod(pod, "")
+	p.PodNominator.AddNominatedPod(framework.NewPodInfo(pod), "")
```
[the remark this answers is not in the document]
Comment (gavinfish) [reply pkg/scheduler/internal/queue/scheduling_queue.go]: done.

File: pkg/scheduler/internal/queue/scheduling_queue.go [hunk 5]
```diff
@@ -324,7 +324,7 @@ func (p *PriorityQueue) AddUnschedulableIfNotPresent(pInfo *framework.QueuedPodI
 		metrics.SchedulerQueueIncomingPods.WithLabelValues("unschedulable", ScheduleAttemptFailure).Inc()
 	}
 
-	p.PodNominator.AddNominatedPod(pod, "")
+	p.PodNominator.AddNominatedPod(framework.NewPodInfo(pod), "")
```
[the remark this answers is not in the document]
Comment (gavinfish) [reply pkg/scheduler/internal/queue/scheduling_queue.go]: done.

File: pkg/scheduler/internal/queue/scheduling_queue.go [hunk 6]
```diff
@@ -793,13 +793,13 @@ func (npm *nominatedPodMap) UpdateNominatedPod(oldPod, newPod *v1.Pod) {
 	// We update irrespective of the nominatedNodeName changed or not, to ensure
 	// that pod pointer is updated.
 	npm.delete(oldPod)
-	npm.add(newPod, nodeName)
+	npm.add(framework.NewPodInfo(newPod), nodeName)
```
[the remark this answers is not in the document]
Comment (gavinfish) [reply pkg/scheduler/internal/queue/scheduling_queue.go]: done.

File: pkg/scheduler/internal/queue/scheduling_queue.go [hunk 7]
```diff
@@ -657,7 +657,7 @@ func (p *PriorityQueue) calculateBackoffDuration(podInfo *framework.QueuedPodInf
 
 func updatePod(oldPodInfo interface{}, newPod *v1.Pod) *framework.QueuedPodInfo {
 	pInfo := oldPodInfo.(*framework.QueuedPodInfo)
-	pInfo.Pod = newPod
+	pInfo.PodInfo = framework.NewPodInfo(newPod)
```
[the remark this answers is not in the document]
Comment (gavinfish) [reply pkg/scheduler/internal/queue/scheduling_queue.go]: done.

File: pkg/scheduler/internal/queue/scheduling_queue.go (hunk 1 above)
Comment (gavinfish) [reply pkg/scheduler/internal/queue/scheduling_queue.go]: done.

File: pkg/scheduler/internal/queue/scheduling_queue.go [hunk 8]
```diff
@@ -412,19 +414,21 @@ func (p *PriorityQueue) Update(oldPod, newPod *v1.Pod) error {
...
 	if oldPod != nil {
-		oldPodInfo := newQueuedPodInfoNoTimestamp(oldPod)
+		oldPodInfo := newQueuedPodInfoForLookup(oldPod)
 		// If the pod is already in the active queue, just update it there.
 		if oldPodInfo, exists, _ := p.activeQ.Get(oldPodInfo); exists {
-			p.PodNominator.UpdateNominatedPod(oldPod, newPod)
-			err := p.activeQ.Update(updatePod(oldPodInfo, newPod))
+			pInfo := updatePod(oldPodInfo, newPod)
+			p.PodNominator.UpdateNominatedPod(oldPod, pInfo.PodInfo)
+			err := p.activeQ.Update(pInfo)
```
Comment (ahg-g) [inline pkg/scheduler/internal/queue/scheduling_queue.go:422]: ```return p.activeQ.Update(pInfo)```

File: pkg/scheduler/internal/queue/scheduling_queue.go [hunk 9]
```diff
@@ -412,19 +414,21 @@ func (p *PriorityQueue) Update(oldPod, newPod *v1.Pod) error {
...
 		}
 
 		// If the pod is in the backoff queue, update it there.
 		if oldPodInfo, exists, _ := p.podBackoffQ.Get(oldPodInfo); exists {
-			p.PodNominator.UpdateNominatedPod(oldPod, newPod)
+			pInfo := updatePod(oldPodInfo, newPod)
+			p.PodNominator.UpdateNominatedPod(oldPod, pInfo.PodInfo)
 			p.podBackoffQ.Delete(oldPodInfo)
-			err := p.activeQ.Add(updatePod(oldPodInfo, newPod))
+			err := p.activeQ.Add(pInfo)
```
Comment (ahg-g) [inline pkg/scheduler/internal/queue/scheduling_queue.go:431]:
```
if err := p.activeQ.Add(pInfo); err != nil {
  return err
}
p.cond.Broadcast()
return nil
```

same thing below please.

File: pkg/scheduler/framework/types.go [hunk 10]
```diff
@@ -79,6 +79,24 @@ type PodInfo struct {
...
+		RequiredAntiAffinityTerms:  pi.RequiredAntiAffinityTerms,
+		PreferredAffinityTerms:     pi.PreferredAffinityTerms,
+		PreferredAntiAffinityTerms: pi.PreferredAntiAffinityTerms,
+		ParseError:                 pi.ParseError,
+	}
+}
+
+// UpdatePodInfo updates fields for PodInfo.
+// PodInfo includes immutable information, and so it is safe to update the pod in place.
+func (pi *PodInfo) UpdatePodInfo(pod *v1.Pod) {
```
[the remark this answers is not in the document]
Comment (gavinfish) [reply pkg/scheduler/framework/types.go]: updated.

File: pkg/scheduler/internal/queue/scheduling_queue.go (hunk 8 above)
Comment (gavinfish) [reply pkg/scheduler/internal/queue/scheduling_queue.go]: done.

File: pkg/scheduler/internal/queue/scheduling_queue.go (hunk 9 above)
Comment (gavinfish) [reply pkg/scheduler/internal/queue/scheduling_queue.go]: done.

File: pkg/scheduler/internal/queue/scheduling_queue.go [hunk 11]
```diff
@@ -412,48 +414,53 @@ func (p *PriorityQueue) Update(oldPod, newPod *v1.Pod) error {
...
 
 	// If the pod is in the unschedulable queue, updating it may make it schedulable.
 	if usPodInfo := p.unschedulableQ.get(newPod); usPodInfo != nil {
-		p.PodNominator.UpdateNominatedPod(oldPod, newPod)
 		if isPodUpdated(oldPod, newPod) {
 			p.unschedulableQ.delete(usPodInfo.Pod)
-			err := p.activeQ.Add(updatePod(usPodInfo, newPod))
+			pInfo := updatePod(usPodInfo, newPod)
+			p.PodNominator.UpdateNominatedPod(oldPod, pInfo.PodInfo)
+			err := p.activeQ.Add(pInfo)
```
Comment (ahg-g) [inline pkg/scheduler/internal/queue/scheduling_queue.go:444]: ditto like the one above

```
if err := p.activeQ.Add(pInfo); err != nil {
  return err
}
p.cond.Broadcast()
return nil
```

File: pkg/scheduler/internal/queue/scheduling_queue.go (hunk 9 above)
Comment (ahg-g) [reply pkg/scheduler/internal/queue/scheduling_queue.go]: there is another one below

**Comment by ahg-g** [comment]
are you sure the kind e2e test failure is not relevant? it seems that the pod is not getting scheduled in those tests.

**Comment by gavinfish** [comment]
> are you sure the kind e2e test failure is not relevant? it seems that the pod is not getting scheduled in those tests.

`pull-kubernetes-e2e-kind` is really unstable recently, referring to [testgrid](https://testgrid.k8s.io/presubmits-kubernetes-blocking#pull-kubernetes-e2e-kind). But yes, this PR may also block it, will keep an eye on it.

File: pkg/scheduler/internal/queue/scheduling_queue.go (hunk 11 above)
Comment (gavinfish) [reply pkg/scheduler/internal/queue/scheduling_queue.go]: done.

**Comment by ahg-g** [comment]
Thanks, pls squash

**Comment by gavinfish** [comment]
squashed

**Comment by ahg-g** [comment]
lets first wait and find a resolution to https://github.com/kubernetes/kubernetes/issues/99304 first.

**Pull request merged**